### PR TITLE
Chemistry Rework: Energy Removal, Recipe Saving

### PR DESCRIPTION
**Pull request by Iamgoofball**
1. >energy
2. this means chemistry can pick up the slack on all the shit that just got nerfed in medbay
3. you can produce basic shit faster with recipe saving

### CODE REVIEW

**Comment by MoreRobustThanYou** [comment]
WOAH WAIT I'LL BE ABLE TO COPY PASTE METH RECIPES? AWESOME

**Comment by kevinz000** [comment]
this is dumb, stop.
don't idlock a crucial machine.

**Comment by Buggy123** [comment]
yes to everything except id lock.

**Comment by DoctoreMad** [comment]
"Yes to everything except HALF OF THE PR"

**Comment by RogueSteampunker** [comment]
ID-locking a crucial machine: Because it wasn't clear enough that you don't play to see what a TERRIBLE IDEA THAT IS

**Comment by kevinz000** [comment]
yeah he doesn't he claims he still knows game design though
even while not knowing how player interaction/round progression goes in actual play because you know
you kinda need to connect to the server and play/observe to see that!

**Comment by calibraptor** [comment]
I personally disagree with nerfing medbay, although I like the idea of being able to copy-paste chemistry recipes for fast fridge stocking.

**Comment by Doppeltongue** [comment]
It it just me or are goof's PRs getting increasingly worse? Do we still have warranty? Can we send him back to the factory or do we need a replacement?

**Comment by Iamgoofball** [comment]
all I'm hearing is
"i cant steal chemistry every round for botany/xenobio/r&d Q_Q"

**Comment by kevinz000** [comment]
@Doppeltongue he's been this way the entire day

**Comment by swindly** [comment]
This locks the booze and soda dispensers to chemists.

**Comment by Cruix** [comment]
:-1: This is a slippery slope, soon we'll be ID locking other job-specific things, like lockers and vending machines. 

Oh wait

**Comment by Doppeltongue** [comment]
@kevinz000 Can you check to see if we still have his receipt? Most stores will only allow customers to exchange for store credit, but maybe we can get a better model.

**Comment by Iamgoofball** [comment]
if you need to use chemistry and you're not a chemist
1. is it lowpop? are you a MD? you have access, lowpop gives you said access
2. are you trying to do it legitimately? go to the HoP.
3. no HoP? if there's a QM they can give you the access from their console
4. none of the above? break into the HoP's office/get the captain
5. none? get the spare
6. spare gone? find who stole it and get them to do it/steal from them

**Comment by Bawhoppen** [comment]
I'd prefer a system where chem dispensers are removed entirely, and they just start with some base ingredient chems, and then have to have them replenished through a combination of mining, botany, cargo, and other departments.

**Comment by Iamgoofball** [comment]
@Bawhoppen bay tried that, went as poorly as you'd expect it to be

**Comment by Bawhoppen** [comment]
Yea but Bay isn't really the best mechanical ss13 experience. I bet it could work here to some extent.

**Comment by RogueSteampunker** [comment]
Well, let's see:

"i cant steal chemistry every round for botany/xenobio/r&d Q_Q"

In reality, it's more "I REALLY don't wanna depend on the idea that Chemistry is actually competent and NOT just sitting there asking how to make meth, or that the HoP hasn't abandoned his position at the second minute mark because he wants to raid the armory"

"is it lowpop? are you a MD? you have access, lowpop gives you said access
are you trying to do it legitimately? go to the HoP.
no HoP? if there's a QM they can give you the access from their console
none of the above? break into the HoP's office/get the captain
none? get the spare
spare gone? find who stole it and get them to do it/steal from them"

1. Valid idea, still worthless in the long run for anyone besides doctors.
2. Once again, the HoP is prone to fucking off the second the initial greytide rolls off from the line. Unreliable at best.
3. Since when can the QM hand out access?
4. So instead of just robbing the Chemists, whose worst consequence is a missing chem machine, rob the people with the shiny access and sensitive items? That seems like a good way to balance things. /s
5. If there's no Captain or HoP to speak of, that spare is gone in the first minute. End of story.
6. By this point, they have the entire armory and will care about nothing but themselves, because it will either be a greytider or the clown. Good luck in that regard.

And let's not forget that now, instead of just taking the machine, people will more than likely mug and/or kill the Chemist for their ID. All you've done is remove the nonlethal aspect, and ensured that SOMEONE is losing more than a single machine.

So, in conclusion, please stop touching things that you really don't seem to understand. The Goof-brand vacuum you're in does not fit the actual gameplay at all.

**Comment by Iamgoofball** [comment]
Mugging/killing the chemist as a non antag is bannable, this is normal

Most players aren't assholes

I meant to say CMO not QM

**Comment by Iamgoofball** [comment]
Also by get the captain I mean "ask the captain to give you access/make you a chemist"

Additionally have you tried asking said chemists nicely

All these problems you're listing are quite frankly just part of the game experience

**Comment by RogueSteampunker** [comment]
"Mugging/killing the chemist as a non antag is bannable, this is normal"

Killing is bannable, mugging will usually go unlooked unless it's unprovoked. And if a Chemist is being stuffy and generally a dick about the machine, it can and will happen. And all adding an ID lock will do is ensure that players are losing more than a single machine.

"Most players aren't assholes"

You see, I'd start laughing, but then that'd take an entire paragraph. So instead, I'm going to point out that a fair few jobs rely on Chemistry for reagents. And if a Chemist decides to be an asshole, places like Science and Botany are suddenly without much for themselves, left to scrape by in hopes of self-sufficiency, which destroys the purpose of cooperation between such departments. So no, not all players are assholes. It just takes one or two assholes in Chemistry to ruin a fair few players.

"Also by get the captain I mean "ask the captain to give you access/make you a chemist""

That works about as well as the HoP. The only difference is that the Captain is probably dead, instead of fucking off. However, a fair point if the Captain isn't busy torturing people in Sec, or pretending to be retarded. (Which is to say never)

"Additionally have you tried asking said chemists nicely"

I have. I've always tried asking first, since Botany can create it's own mutagen with Blumpkins. However, I have been met with more silence and harsh words rather than actual assistance. There have been amazing Chemists, yes, but moreso the sort that either attempt to ignore me, or tell me that they're doing other things. Those other things, of course, being them creating floor pills full of mindbreaker toxin. I never resort to stealing unless necessary, and Chemists are enough of a mixbag that ID locking chem machines will lead to a high amount of Chemistry breakins.

"All these problems you're listing are quite frankly just part of the game experience"

So dealing with people that stop me from doing my job is game experience? Having to jump through hoop after hoop so I can do a simple task is game experience? If I wanted to take 50 minutes to do a simple action, I'd go play on a HRP server rather than go through a convoluted set of instructions if the Chemist is more focused on memeing than actually doing their job.

**Comment by Iamgoofball** [comment]
i think you're freaking out over an issue that will come up like, 1 in every 1000 rounds(except the first week people will drum it up to complain)

i'm also opening 2 more chemist slots with the next update to this PR, and i think it'll be ok that under some circumstances, you can't use the chemistry machine

the communications console/security console/medical records/access console aren't publicly usable, but people aren't clamoring to unlock those

**Comment by ExcessiveUseOfCobblestone** [comment]
> the communications console/security console/medical records/access console aren't publicly usable, but people aren't clamoring to unlock those

We really should have a way to hack those.

**Comment by Iamgoofball** [comment]
you can emag them all the same

**Comment by Bawhoppen** [comment]
We shouldn't encourage making emags stronger

**Comment by RogueSteampunker** [comment]
"the communications console/security console/medical records/access console aren't publicly usable, but people aren't clamoring to unlock those"

Because those don't affect most jobs to as much of a degree as a chem machine. I couldn't give a damn if I can't use a security console if I'm not security, but I'll certainly give one if I'm a Scientist or Botanist without access to a chem machine if needed.

And honestly, the sort of situations detailed about aren't rare. They don't happen every round, but they happen enough that ID locking the dispensers can and will cause issues. And that's only covering round-start needs, and not when the antics and general situations of the round start rolling. I may be a little exaggerated, but you're seriously downplaying the idea of how important chem machines being usable by all is.

**Comment by Iamgoofball** [comment]
sometimes you can't get what you want

**Comment by Iamgoofball** [comment]
additionally if you're a scientist/botanist, and all 6 potential methods to get access to the machine aren't available, you've got bigger  problems then printing boards or making plants grow weird

**Comment by RogueSteampunker** [comment]
Goof, I feel that all of your argument boils down to the idea that you think anything that stops a player from actually doing their role is "player experience". And that's just sad.

"additionally if you're a scientist/botanist, and all 6 potential methods to get access to the machine aren't available, you've got bigger problems then printing boards or making plants grow weird"

Last I checked, 1 through 3 are what normally happens with heads, and 4 through 6 are validhunting the people with the access. At worst, a head or two got knocked. At it's most likely, the Heads are doing their things, and chasing the greytide with a taser is a lost cause.

"sometimes you can't get what you want"

That is literally the most lazy justification for a change I've ever heard for a change that's literally just "Muh machines keep being stolen cause I'm a cunt".

**Comment by Iamgoofball** [comment]
quite frankly, if our playerbase is so shit that basic teamwork is so impossible and everything has to be able to be done solo otherwise everything falls apart

we've got bigger problems to deal with

**Comment by PKPenguin321** [comment]
Stupid :-1:

**Comment by Iamgoofball** [comment]
@PKPenguin321 tough titties, blame the players who greytided chemistry/threatened to greytide it over other balance changes for this

**Comment by PKPenguin321** [comment]
I'm blaming you for your asinine solution to an imaginary problem

**Comment by Shadowlight213** [comment]
\>doesn't even compile

**Comment by Iamgoofball** [comment]
have a fix locally, working on >>>tgui

**Comment by Qustinnus** [comment]
recipe saving is cool. this is literaly just being shit on because its a goof PR at this point it seems like.

**Comment by Iamgoofball** [comment]
fuck
fuck
fuck

**Comment by TribeOfBeavers** [comment]
What are the resteictions on saving a recipe? If there aren't any chemistry just becomes copy/pasting the recipe off the wiki or forum or whatever then spamming an unlimited amount of it with no effort at all.

**Comment by Iamgoofball** [comment]
You can only punch in already dispensable reagents so this just helps you avoid carpal tunnel syndrome for the basic shit like mutagen you get bugged about all the time

**Comment by Iamgoofball** [comment]
And you can't heat reagents from the dispenser at all

**Comment by Grazyn** [comment]
Note that stealing the dispenser had become so frequent and cancerous at some point, that it's now used as the case study in every discussion about escalation. Just think how many policy threads and bans and notes could've been avoided if the dispenser had been ID locked from the start.

**Comment by Iamgoofball** [comment]
I like to firebomb botany when they steal the hallway dispenser

**Comment by deathride58** [comment]
With how important the chem dispenser is, especially when antags inevitably start murderboning or when medbay is inevitably incompetent, locking it behind an ID access requirement is a pretty dumb idea.

If you really want to solve the problem of assistants jacking the chem dispenser while solving numerous other problems, just make major chem recipes randomized every round, and give chemistry's chemmaster's analysis function the ability to see what chems can be made from a reagent, and what chems go towards making a reagent. Alternatively, instead of the chemmasters receiving a buff, chemists and the CMO could start with a chemistry cheatsheet listing the randomized recipes, similar to how the station blueprints display the wires for machinery in a given round. With either solution, it'd still be possible to pump out healing chems from the chem dispensers when all of medbay died in a maxcap, but the strategy of stealing a chem dispenser for unstable mutagen would receive a huge nerf due to the recipe being random every round.

**Comment by Iamgoofball** [comment]
>just make major chem recipes randomized every round

I'm not recoding our chemistry system to fix what an ID lock fixes

**Comment by Arianya** [comment]
Alternatively just make them like consoles, requiring more then just a wrench to relocate

**Comment by deathride58** [comment]
The bare minimum you'd have to do in order to make chem recipes random?
Move the reactant lists of various chems to global lists (or shove em in their own dedicated subsystem), then write up proc that randomizes the contents of those lists using the roundstart chem dispenser's chems as a pool to pull from. It should only take around ten minutes of writing code to have a working randomized chem recipe system, not counting the amount of time needed to compile and test.

Increasing the number of steps required to steal a chem dispenser is also a pretty good potential solution to the issue, and would help differentiate the roundstart dispensers from lateround science's portable chem dispensers.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 1]
```diff
@@ -157,24 +147,50 @@
...
+						return
+				saved_recipes += list(list("recipe_name" = name, "contents" = recipe))
 		if("eject")
 			if(beaker)
 				beaker.forceMove(loc)
 				beaker = null
 				cut_overlays()
 				. = TRUE
 
+/obj/machinery/chem_dispenser/proc/process_recipe_list(var/fucking_hell)
```
Comment (RemieRichards) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: This form actually already exists, Byond Param strings, as such ```params2list()``` can replace this proc.

File: tgui/src/interfaces/chem_dispenser.ract [hunk 2]
```diff
@@ -15,6 +10,14 @@
     {{/each}}
   </ui-section>
 </ui-display>
+<ui-display title='Saved Recipes' button>
+  <ui-section>
+    <ui-button icon='plus' action='add_recipe'>Add Recipe</ui-button>
```
Comment (RemieRichards) [inline tgui/src/interfaces/chem_dispenser.ract]: this would look better on its own line, above the grid.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 3]
```diff
@@ -9,11 +9,8 @@
 	idle_power_usage = 40
 	interact_offline = 1
 	resistance_flags = FIRE_PROOF | ACID_PROOF
-	var/cell_type = /obj/item/stock_parts/cell/high
-	var/obj/item/stock_parts/cell/cell
-	var/powerefficiency = 0.01
+	req_access = (ACCESS_CHEMISTRY)
```
Comment (RemieRichards) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: ```req_access = list(ACCESS_CHEMISTRY)```

**Comment by The-clown-known-as-Pepe** [comment]
make sure that snowflake chem dispenser that's on metastation xenobio has access for scientists

**Comment by CPTANT** [comment]
ID locking machines is lame.

Rooms should be locked, machines free to use, save for the most high security things.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: Perfect

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 3 above)
Comment (Iamgoofball) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: Whoops

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: Wait, does it support the delimiters I use

**Comment by Qbopper** [comment]
ITT: goof tries to "rework" medbay by removing shit he doesn't like and leaving the actual flaws with the role in because that's too much effort to code

in fact i'm 99% certain the reason goof said he wouldn't code that rework proposal he helped make was something to do with how "it would just get rejected anyways"

really makes you think

**Comment by Iamgoofball** [comment]
@deathride58

Not only is that not how the chemistry backend works, the amount of overhead to ensure recipes dont override other recipes due to matching recipes will be insane. It's not happening.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: Negative, this won't work for me. I tried params2list with my format, it doesn't work.

**Comment by uraniummeltdown** [comment]
ID locks are lame, this isn't Bay
Recipe saving thing is cool tho 👍

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 1 above)
Comment (RemieRichards) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: it's ```"name1=value1&name2=value2&..."```, just swap your ; for &'s, and & is more natural for a user anyway.

**Comment by Hatfish** [comment]
When will we get an ID lock on R&D and engineering equipment then?  If you want to clamp down on moonlighters you may as well go all the way.

**Comment by Militaires** [comment]
>When will we get an ID lock on R&D and engineering equipment then? If you want to clamp down on moonlighters you may as well go all the way.

this already exists it's called locking the console

**Comment by Qbopper** [comment]
adding an id lock to chemistry machines is a bad idea because it means botanists won't steal the console anymore, how the fuck will i get my valids then

**Comment by Hatfish** [comment]
Locking the console isn't the same as this, you can choose to leave that unlocked.  This is an ID scan, so you don't even have the option.

**Comment by FantasticFwoosh** [comment]
Just have chemistry energy ports attached to the floor or something to justify no internal battery supply @Iamgoofball , move the chem machine off the port & poof, it wont connect to wires or local rooms unless its put back. Alternative is to make a shitty but up-gradable battery portable version.

I dont care about id-locking its a salt-change thats going to happen anyway if nothing radical changes in the machines usage, i think this PR's outlying changes are bad though because it acts like lack of waiting literally two seconds for a recharge (a system adopted by other machines linked to the capacitors to make it faster) is actually a hindering problem.

**Comment by Bawhoppen** [comment]
ID locking is also arbitrary mechanic that we shouldn't build more stuff onto. ID locking the chem dispenser is only really a bandaid solution honestly, where we really need a more dynamic solution to this.

**Comment by Iamgoofball** [comment]
Fwoosh, these changes are needed. Go home.

**Comment by Qbopper** [comment]
\>admits to the fact that he barely plays the game
\>feels he knows enough about the game's state to say that these changes are needed

goof pls

**Comment by FantasticFwoosh** [comment]
Reading was never this community's strong suit.

Besides, any rogue silicon force can just murder the chemists & remotely ID lock the dispensers in order to stop grenade production.

**Comment by Iamgoofball** [comment]
What are you rambling about

**Comment by kevinz000** [comment]
@Militaires you're not realizing the fact that console locks can easily be circumvented. 
just stop goofball you don't even play the game why are you trying to make it into your personal vision?

**Comment by FantasticFwoosh** [comment]
@Iamgoofball The AI or borgs will probably turn off & lock meatbags out because they have access to chem machines and bypass the ID lock in response to any chemical misuse or threat for particularly safety obnoxious silicons. This extends to blocking attempts at making grenades vs blobs if crewmembers blow each other up or to counter specific threats.

I thought i was clear the first time. You really didn't know that the AI/borgs can also manually lock & unlock all non-sign in ID systems (can't access HOP console unless HOP's ID is already inserted to click the button & sign in or if its already logged in.) like the RnD consoles lock screen?

**Comment by KorPhaeron** [comment]
I don't like ID locks (the exceptions being on the command consoles/HoP console/the computer that lets you remotely delete players and I hope its obvious why those are exceptions)

**Comment by KorPhaeron** [comment]
Recipe stuff is cool!

**Comment by Qustinnus** [comment]
>18 hours

**Comment by KorPhaeron** [comment]
I forgot I can just

**Comment by Iamgoofball** [comment]
Perfect

**Comment by Buggy123** [comment]
Oh, i was gonna make a new PR that just removes the ID lock but that works too 👍

**Comment by epochayur** [comment]
with the removal of the ID lock this is a halfway decent PR

**Comment by Buggy123** [comment]
Yeah, 100% support the pr with no ID lock. Energy is a annoying limitation especially since the chem dispenser can't be upgraded, and chem recipes are a much-needed feature.

**Comment by uraniummeltdown** [comment]
Why don't we just make regular chem dispensers have stock parts in them?
Infinite energy is retarded, being able to upgrade the capacitor is better

**Comment by RogueSteampunker** [comment]
Forgot to mention, I do actually like the rest of this. I'm in full support of giving chemists an easier time, just not an exclusive right

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: the thing is the docs say it should also support ;, which looks nicer

i'll give it a shot I guess

**Comment by Qbopper** [comment]
recipes are saved on a per round basis to the machine, right?

**Comment by Iamgoofball** [comment]
@Qbopper yeah

**Comment by Qbopper** [comment]
cool idea

I don't think energy has ever been relevant either so yeah not an awful pr anymore

**Comment by Jalleo** [comment]
Make sure the dispeners dont work if the room isn't powered please.

**Comment by Iamgoofball** [comment]
@Jalleo they already dont like any other machine

**Comment by newfren** [comment]
@Iamgoofball currently the dispensers still function even if chemistry is out of power by drawing on its internal energy supply. This means if chem runs out of power you can still mix styptic/silver sulf and splash it on people or whatever. This is what Jalleo is referring to.

**Comment by Iamgoofball** [comment]
hm

**Comment by ExcessiveUseOfCobblestone** [comment]
That's because the chem machine's power/energy is based off an internal powercell rather than the APC.

**Comment by Iamgoofball** [comment]
Internal cells were ditched.

**Comment by ExcessiveUseOfCobblestone** [comment]
HMMMM

**Comment by Jcenderion** [comment]
Why not make it a 'half' ID lock. Make it so the dispenser can only make healing chems without an ID but with an ID all recipes can be made. I feel like that's a good compromise, albeit might be somewhat difficult.

**Comment by Qustinnus** [comment]
@Rinoka 
>"invest the effort or dont do it"
>doesn't even contribute to the codebase themselves

go away, the id lock is already gone and re-doing the entire chemistry system is an ass load of work.

**Comment by PKPenguin321** [comment]
@Qustinnus you must admit that goofball isn't known for extensive or prolonged effort though

**Comment by Iamgoofball** [comment]
pk ur an awfy hunan being

**Comment by Magikarpet** [comment]
golfball
one round in a single button press with no cooldown that ignroes the other half of the log spam it can commit
